tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,426 @@
+        #  mutated_view_idx is only needed for lambdas of multi-output view ops, like split().
+        # For the forward lambda, we just directly call the corresponding view op and index into the output,
+        # which is why mutated_view_idx isn't actually an argument to the function call.
+        # For the reverse lambda, we have a calling convention for every {view}_reverse function such that
+        # it only accepts an additional mutated_view_idx parameter if it's a multi-output view op.
+        maybe_idx = f'[{forward_lambda_exprs[-1]}]'
+        forward_lambda_exprs = forward_lambda_exprs[:-1]
+
+    if Variant.method in f.variants:
+        forward_lambda_call_str = f'{forward_lambda_exprs[0]}.{forward_name}({", ".join(forward_lambda_exprs[1:])}){maybe_idx}'
+    else:
+        forward_lambda_call_str = f'at::{forward_name}({", ".join(forward_lambda_exprs)}){maybe_idx}'
+
+    view_meta_str = f"""
+      at::functionalization::ViewMeta view_meta = at::functionalization::ViewMeta(
+        [{captured_view_args_str}]({', '.join(a.decl() for a in forward_lambda_args)}) -> {forward_ret_type.cpp_type()} {{
+          return {forward_lambda_call_str};
+        }},
+        [{captured_view_args_str}]({', '.join(a.decl() for a in reverse_lambda_args)}) -> {reverse_ret_type.cpp_type()} {{
+          return at::functionalization::impl::{reverse_name}({", ".join(reverse_lambda_exprs)});
+        }}
+      );"""
+
+    api_name = CppSignatureGroup.from_native_function(
+        f, method=False, fallback_binding=f.manual_cpp_binding).most_faithful_signature().name()
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # This op is both an inplace op AND a view op.
+        # See Note [Functionalization Pass - Inplace View Ops] for details.
+        # I currently have the view meta call into the out-of-place variant of the view, to avoid
+        # having to define an extra ~20 inplace {view}_inverse_ functions.
+        # Most view ops don't have NativeFunctionGroup's both, because we don't define out= variants for view ops.
+        # I'm assuming that every inplace-view op has a corresponding out-of-place view op,
+        # with the same name but the trailing underscore removed.
+        # This is currently asserted at parse time in gen.py (see error_check_native_functions).
+        assert f.func.kind() is SchemaKind.inplace
+        api_name = api_name[:-1]
+
+    # The meta API call should use the same arguments, but convert all tensors to meta tensors first.
+    meta_conversion_str, meta_call_ctx = convert_to_meta_tensors(dispatcher_sig)
+    meta_call_args = [e.expr for e in translate(meta_call_ctx, dispatcher_sig.arguments(), method=False)]
+    if Variant.method in f.variants:
+        assert is_tensor_like(meta_call_ctx[0].argument)
+        meta_call_str = f'{meta_call_ctx[0].name}.{api_name}({", ".join(meta_call_args[1:])})'
+    else:
+        meta_call_str = f'at::{api_name}({", ".join(meta_call_args)})'
```

Comment:
It is incredibly strange to me that you call api_name here (from CppSignatureGroup), but meta_call_args is translated from `dispatcher_sig.arguments()`. API confusion!

Reply:
yup, I think I was just lucky here because the main difference in the arguments between the two API's is TensorOptions, and the codegen here only cares about view/inplace ops (which don't have TensorOptions args).

Just laying it all out:
- `at::`, `at::Tensor::`, and `at::redispatch::` are all frontend API's that follow the C++ API (used in `CppSignature`)
- I remembered that enough to use the `CppSignature` code to correctly get the name of the target function to call, but, I'm incorrectly using the dispatcher API to bind arguments here.
- What I should really do instead is ignore the C++ API completely, and use the `at::_ops` API, that way I can stick entirely in dispatcher API land (since the input arguments to my kernels follow the dispatcher API)